components/eamxx/src/share/field/field_utils_impl.hpp

```diff
@@ -374,10 +429,53 @@ void horiz_contraction(const Field &f_out, const Field &f_in,
     comm->all_reduce(f_out.template get_internal_view_data<ST, Host>(),
                      l_out.size(), MPI_SUM);
     f_out.sync_to_dev();
+    if (is_comm_avg_masked) {
+      f_tmp.sync_to_host();
+      comm->all_reduce(f_tmp.template get_internal_view_data<ST, Host>(),
```

Comment:
Be aware that `get_internal_view_data` is a power-user method. In particular, it is not safe to use with fields that are subfields of another field (where their entries may not be contiguous in memory). Maybe you could add a check at the fcn top (among all others checks on f_tmp) to ensure it's not a subfield?

TBC, I don't think anyone would ever pass a subfield as a scratch, but better be safe than sorry..

Reply:
To be precise, we _can_ support subfields as long as they are subfields along the slowest striding dim, since once we extract the view, we get contiguous data, but it would require a few if/else (basically extract the view, then take the view pointer). The method `get_internal_view_data` does not do "get view, then take the view pointer", but it simply returns the internal view pointer, which, for subfields, is the pointer to the global "super"-field. Boring impl details.